File: solutions/day-40-permutation-in-string/Solution.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        if (s1.length() > s2.length()) return false;

        int[] need = new int[26];
        int[] window = new int[26];

        for (char c : s1.toCharArray()) {
            need[c - 'a']++;
        }

        for (int i = 0; i < s2.length(); i++) {
            window[s2.charAt(i) - 'a']++;

            if (i >= s1.length()) {
                window[s2.charAt(i - s1.length()) - 'a']--;
            }

            if (java.util.Arrays.equals(window, need)) {
                return true;
            }
        }

        return false;
    }
}

File: solutions/day-40-permutation-in-string/Solution.cpp
#include <string>
#include <vector>

class Solution {
public:
    bool checkInclusion(std::string s1, std::string s2) {
        if (s1.size() > s2.size()) return false;

        std::vector<int> need(26, 0);
        std::vector<int> window(26, 0);

        for (char c : s1) {
            need[c - 'a']++;
        }

        for (int i = 0; i < static_cast<int>(s2.size()); ++i) {
            window[s2[i] - 'a']++;

            if (i >= static_cast<int>(s1.size())) {
                window[s2[i - s1.size()] - 'a']--;
            }

            if (window == need) {
                return true;
            }
        }

        return false;
    }
};
